File: Go/MonteCarloSimulationFinance/MonteCarloSimulationFinance.go
package main

import (
	"fmt"
	"math"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/chobie/go-gaussian"
	"github.com/markcheno/go-quote"
	"gonum.org/v1/gonum/stat"
)

// MonteCarloSimulationFinance is a structure to predict stock market prices using Monte Carlo simulation.
type MonteCarloSimulationFinance struct {
	numberOfProcesses int
	timeSeries        []float64
	startDate         string
	endDate           string
	tickerSymbol      string
	data              []float64
}

// https://github.com/markcheno/go-quote
// https://godoc.org/github.com/markcheno/go-quote#NewQuoteFromYahoo
func (monteCarloSimulationFinance *MonteCarloSimulationFinance) dataAcquisition() {
	stock, _ := quote.NewQuoteFromYahoo(monteCarloSimulationFinance.tickerSymbol,
		monteCarloSimulationFinance.startDate, monteCarloSimulationFinance.endDate, quote.Daily, true)
	monteCarloSimulationFinance.timeSeries = stock.Close
}

/*  A z-table, also called the standard normal table, is a mathematical table that allows us to know
the percentage of values below (to the left) a z-score in a standard normal distribution (SND).
A z-score, also known as a standard score, indicates the number of standard deviations
a raw score lays above or below the mean. When the mean of the z-score is calculated it is always 0,
and the standard deviation (variance) is always in increments of 1. */
func (monteCarloSimulationFinance *MonteCarloSimulationFinance) calculateZScore() float64 {
	normalDistribution := gaussian.NewGaussian(0, 1)
	/* 	Source code for random number generator https://play.golang.org/p/ZdFpbahgC1
	   	The default number generator is deterministic, so it'll
	   	produce the same sequence of numbers each time by default.
	   	To produce varying sequences, give it a seed that changes.
	   	Note that this is not safe to use for random numbers you
	   	intend to be secret, use `crypto/rand` for those.
	   	Seeding - Go provides a method, Seed(seed int64), that allows you
	   	to initialize this default sequence. Implementation is slow
	   	to make it faster rand.Seed(time.Now().UnixNano()) is added.
	   	Seed is the current time, converted to int64 by UnixNano.
	   	Gives constantly changing numbers. */

	// Seed
	s := rand.NewSource(time.Now().UnixNano())
	// Randomly changing numbers.
	r := rand.New(s)
	/* 	Call the resulting `rand.Rand` just like the
	   	functions on the `rand` package. */
	return normalDistribution.Ppf(r.Float64())

}

// Differencing time series = Shifting and lagging time series
func (monteCarloSimulationFinance *MonteCarloSimulationFinance) calculatePeriodicDailyReturn() {
	for i := 1; i < len(monteCarloSimulationFinance.timeSeries); i++ {
		monteCarloSimulationFinance.data = append(monteCarloSimulationFinance.data,
			math.Log(monteCarloSimulationFinance.timeSeries[i]/monteCarloSimulationFinance.timeSeries[i-1]))
	}
}

// https://github.com/gonum/gonum
// https://godoc.org/gonum.org/v1/gonum/stat#Mean
func (monteCarloSimulationFinance *MonteCarloSimulationFinance) calculateAverageDailyReturn() float64 {
	/* 	computes the weighted mean of the dataset.
	   	we don't have any weights (ie: all weights are 1)
	   	so we just pass a nil slice. */
	return stat.Mean(monteCarloSimulationFinance.data, nil)
}

// https://github.com/gonum/gonum
// https://godoc.org/gonum.org/v1/gonum/stat#Variance
func (monteCarloSimulationFinance *MonteCarloSimulationFinance) calculateVariance() float64 {
	/* 	computes the weighted variance of the dataset.
	   	we don't have any weights (ie: all weights are 1)
	   	so we just pass a nil slice. */
	return stat.Variance(monteCarloSimulationFinance.data, nil)

}

func (monteCarloSimulationFinance *MonteCarloSimulationFinance) calculateStandardDeviation() float64 {
	return math.Sqrt(monteCarloSimulationFinance.calculateVariance())

}

func (monteCarloSimulationFinance *MonteCarloSimulationFinance) calculateDrift() float64 {
	return monteCarloSimulationFinance.calculateAverageDailyReturn() - monteCarloSimulationFinance.calculateVariance()/2
}

func (monteCarloSimulationFinance *MonteCarloSimulationFinance) calculateRandomValue() float64 {
	return monteCarloSimulationFinance.calculateStandardDeviation() * monteCarloSimulationFinance.calculateZScore()
}

func (monteCarloSimulationFinance *MonteCarloSimulationFinance) simulationFinance(numberOfSimulations int, predictionWindowSize int, channel chan [][]float64) {
	var prediction []float64
	var predictions [][]float64
	for i := 0; i < numberOfSimulations; i++ {
		// today’s price
		prediction = append(prediction, monteCarloSimulationFinance.timeSeries[len(monteCarloSimulationFinance.timeSeries)-1])
		for j := 0; j < predictionWindowSize; j++ {
			// Next Day’s Price=Today’s Price × e^(Drift+Random Value)
			prediction = append(prediction,
				prediction[len(prediction)-1]*math.Pow(math.E, (monteCarloSimulationFinance.calculateDrift()+monteCarloSimulationFinance.calculateRandomValue())))
		}
		predictions = append(predictions, prediction)
		/* 	Setting the slice to nil is the best way to clear a slice.
		   	nil slices in go are perfectly well behaved and setting the slice to nil
		   	will release the underlying memory to the garbage collector. */
		prediction = nil
	}

	channel <- predictions

}

func (monteCarloSimulationFinance *MonteCarloSimulationFinance) mcsFinanceSerial(numberOfSimulations int, predictionWindowSize int) ([][]float64, float64) {
	startTime := time.Now()
	channel := make(chan [][]float64)
	go monteCarloSimulationFinance.simulationFinance(numberOfSimulations, predictionWindowSize, channel)
	predictions := <-channel
	executionTime := time.Since(startTime).Seconds()
	return predictions, executionTime
}

func (monteCarloSimulationFinance *MonteCarloSimulationFinance) mcsFinanceParallel(numberOfSimulations int, predictionWindowSize int) ([][][]float64, float64) {
	startTime := time.Now()
	numberOfSimulationsPerProcess := numberOfSimulations / monteCarloSimulationFinance.numberOfProcesses
	/* 	Buffered channels are useful when you know how many goroutines you have launched,
	   	want to limit the number of goroutines you will launch, or want to limit
	   	the amount of work that is queued up. */
	channel := make(chan [][]float64, monteCarloSimulationFinance.numberOfProcesses)
	for i := 0; i < monteCarloSimulationFinance.numberOfProcesses; i++ {
		go monteCarloSimulationFinance.simulationFinance(numberOfSimulationsPerProcess, predictionWindowSize, channel)
	}

	var predictions [][][]float64
	for i := 0; i < monteCarloSimulationFinance.numberOfProcesses; i++ {
		prediction := <-channel
		predictions = append(predictions, prediction)

	}
	executionTime := time.Since(startTime).Seconds()
	return predictions, executionTime
}

func (monteCarloSimulationFinance *MonteCarloSimulationFinance) exportFinanceFileSerial(serialPredictions [][]float64) {
	//One row in the output file.
	var row strings.Builder
	var serialNumber int64 = 1
	var predictions [][]float64 = serialPredictions
	f, err := os.Create("C:\\Users\\Dule\\Desktop\\NAPREDNE TEHNIKE PROGRAMIRANJA\\PROJEKAT\\NTP\\Execution Results\\Finance\\GolangFinanceSerial.txt") // creating...
	if err != nil {
		fmt.Printf("Error while creating a file: %v", err)
		return
	}
	defer f.Close()
	for i := 0; i < len(predictions); i++ {
		for j := 0; j < len(predictions[i]); j++ {
			if j == 0 {
				// Write a serial number of simulations at the beginning of a file.
				var serialNumberOfSimulation string = strconv.FormatInt(serialNumber, 10)
				row.WriteString(serialNumberOfSimulation + "," + " ")
			}

			//A component in one row of the output file.
			var component string = strconv.FormatFloat(predictions[i][j], 'f', 7, 64)
			row.WriteString(component)
			if j == (len(predictions[i]) - 1) {
				row.WriteString("\r\n")
				serialNumber++
			} else {
				row.WriteString("," + " ")
			}
		}
	}
	// Writing to file
	_, err = f.WriteString(row.String())
	if err != nil {
		fmt.Printf("Error while writing a file: %v", err)
	}

}

func (monteCarloSimulationFinance *MonteCarloSimulationFinance) exportFinanceFileParallel(parallelPredictions [][][]float64) {
	//One row in the output file.
	var row strings.Builder
	var serialNumber int64 = 1
	var predictions [][][]float64 = parallelPredictions
	// Creating file
	f, err := os.Create("C:\\Users\\Dule\\Desktop\\NAPREDNE TEHNIKE PROGRAMIRANJA\\PROJEKAT\\NTP\\Execution Results\\Finance\\GolangFinanceParallel.txt")
	if err != nil {
		fmt.Printf("Error while creating a file: %v", err)
		return
	}
	defer f.Close()
	for i := 0; i < len(predictions); i++ {
		for j := 0; j < len(predictions[i]); j++ {
			for k := 0; k < len(predictions[i][j]); k++ {
				if k == 0 {
					// Write a serial number of simulations at the beginning of a file.
					var serialNumberOfSimulation string = strconv.FormatInt(serialNumber, 10)
					row.WriteString(serialNumberOfSimulation + "," + " ")
				}

				//A component in one row of the output file.
				var component string = strconv.FormatFloat(predictions[i][j][k], 'f', 7, 64)
				row.WriteString(component)
				if k == (len(predictions[i][j]) - 1) {
					row.WriteString("\r\n")
					serialNumber++
				} else {
					row.WriteString("," + " ")
				}

			}
		}
	}
	// Writing to file
	_, err = f.WriteString(row.String())
	if err != nil {
		fmt.Printf("Error while writing a file: %v", err)
	}

}

func exportScalingFile(scaling string, strongFlag bool) {
	var path string
	if strongFlag == true {
		path = "C:\\Users\\Dule\\Desktop\\NAPREDNE TEHNIKE PROGRAMIRANJA\\PROJEKAT\\NTP\\Scaling Results\\Finance\\GolangPiStrongScaling.csv"
	} else {
		path = "C:\\Users\\Dule\\Desktop\\NAPREDNE TEHNIKE PROGRAMIRANJA\\PROJEKAT\\NTP\\Scaling Results\\Finance\\GolangPiWeakScaling.csv"
	}
	f, err := os.Create(path) // creating...
	if err != nil {
		fmt.Printf("Error while creating a file: %v", err)
		return
	}
	defer f.Close()
	// Writing to file
	_, err = f.WriteString(scaling)
	if err != nil {
		fmt.Printf("Error while writing a file: %v", err)
	}

}

var serialPart float64 = 0.0
var parallelPart float64 = 1.0

func calculateAmdahlSpeedup(numberOfProcesses int) float64 {
	return 1.0 / (serialPart + parallelPart/float64(numberOfProcesses))
}

func calculateGustafsonSpeedup(numberOfProcesses int) float64 {
	return serialPart + parallelPart*float64(numberOfProcesses)
}

func strongScaling() {
	fmt.Println("=======================")
	fmt.Println("Start strong scaling:")
	fmt.Println("=======================")
	//One row in the output file.
	var row strings.Builder
	row.WriteString("number_of_processes,achieved_speedup,theoretical_maximum_speedup\r\n")
	numberOfSimulations := 100
	predictionWindowSize := 10
	numberOfProcessesSerial := 1
	monteCarloSimulationFinanceSerial := MonteCarloSimulationFinance{
		tickerSymbol:      "AAPL",
		startDate:         "2000-01-01",
		endDate:           "2020-01-01",
		numberOfProcesses: numberOfProcessesSerial}
	monteCarloSimulationFinanceSerial.dataAcquisition()
	monteCarloSimulationFinanceSerial.calculatePeriodicDailyReturn()
	fmt.Println("Stock market price predictions using the Monte Carlo simulation serial version")
	_, serialExecutionTime := monteCarloSimulationFinanceSerial.mcsFinanceSerial(numberOfSimulations, predictionWindowSize)
	fmt.Printf("Execution time(n = %d, p = %d, w = %d) = %f seconds\r\n",
		numberOfSimulations, numberOfProcessesSerial,
		predictionWindowSize, serialExecutionTime)

	for numberOfProcessesParallel := 2; numberOfProcessesParallel < 14; numberOfProcessesParallel++ {

		monteCarloSimulationFinanceParallel := MonteCarloSimulationFinance{
			tickerSymbol:      "AAPL",
			startDate:         "2000-01-01",
			endDate:           "2020-01-01",
			numberOfProcesses: numberOfProcessesParallel}
		monteCarloSimulationFinanceParallel.dataAcquisition()
		monteCarloSimulationFinanceParallel.calculatePeriodicDailyReturn()
		fmt.Println("Stock market price predictions using the Monte Carlo simulation parallel version")
		_, parallelExecutionTime := monteCarloSimulationFinanceParallel.mcsFinanceParallel(numberOfSimulations, predictionWindowSize)

		fmt.Printf("Execution time(n = %d, p = %d, w = %d) = %f seconds\r\n",
			numberOfSimulations, numberOfProcessesParallel,
			predictionWindowSize, parallelExecutionTime)

		achievedSpeedup := serialExecutionTime / parallelExecutionTime
		theoreticalMaximumSpeedup := calculateAmdahlSpeedup(numberOfProcessesParallel)
		fmt.Printf("Achieved speedup is: %f times.\r\n", achievedSpeedup)
		fmt.Printf("Theoretical maximum speedup according to Amdahl’s law is: %f times.\r\n", theoreticalMaximumSpeedup)
		row.WriteString(strconv.FormatInt(int64(numberOfProcessesParallel), 10) + "," + strconv.FormatFloat(achievedSpeedup, 'f', -1, 64) + "," + strconv.FormatInt(int64(theoreticalMaximumSpeedup), 10) + "\r\n")
	}
	exportScalingFile(row.String(), true)
	fmt.Println("End strong scaling.")

}
func weakScaling() {
	fmt.Println("=======================")
	fmt.Println("Start weak scaling:")
	fmt.Println("=======================")
	//One row in the output file.
	var row strings.Builder
	row.WriteString("number_of_processes,achieved_speedup,theoretical_maximum_speedup\r\n")
	numberOfSimulations := 100
	predictionWindowSize := 10
	for numberOfProcesses := 2; numberOfProcesses < 14; numberOfProcesses++ {
		increasedNumberOfSimulations := numberOfSimulations * numberOfProcesses
		monteCarloSimulationFinanceSerial := MonteCarloSimulationFinance{
			tickerSymbol:      "AAPL",
			startDate:         "2000-01-01",
			endDate:           "2020-01-01",
			numberOfProcesses: numberOfProcesses}
		monteCarloSimulationFinanceSerial.dataAcquisition()
		monteCarloSimulationFinanceSerial.calculatePeriodicDailyReturn()
		fmt.Println("Stock market price predictions using the Monte Carlo simulation serial version")
		_, serialExecutionTime := monteCarloSimulationFinanceSerial.mcsFinanceSerial(increasedNumberOfSimulations, predictionWindowSize)
		fmt.Printf("Execution time(n = %d, p = %d, w = %d) = %f seconds\r\n",
			increasedNumberOfSimulations, numberOfProcesses,
			predictionWindowSize, serialExecutionTime)

		monteCarloSimulationFinanceParallel := MonteCarloSimulationFinance{
			tickerSymbol:      "AAPL",
			startDate:         "2000-01-01",
			endDate:           "2020-01-01",
			numberOfProcesses: numberOfProcesses}
		monteCarloSimulationFinanceParallel.dataAcquisition()
		monteCarloSimulationFinanceParallel.calculatePeriodicDailyReturn()
		fmt.Println("Stock market price predictions using the Monte Carlo simulation parallel version")
		_, parallelExecutionTime := monteCarloSimulationFinanceParallel.mcsFinanceParallel(increasedNumberOfSimulations, predictionWindowSize)

		fmt.Printf("Execution time(n = %d, p = %d, w = %d) = %f seconds\r\n",
			increasedNumberOfSimulations, numberOfProcesses,
			predictionWindowSize, parallelExecutionTime)

		achievedSpeedup := serialExecutionTime / parallelExecutionTime
		theoreticalMaximumSpeedup := calculateAmdahlSpeedup(numberOfProcesses)
		fmt.Printf("Achieved speedup is: %f times.\r\n", achievedSpeedup)
		fmt.Printf("Theoretical maximum speedup according to Amdahl’s law is: %f times.\r\n", theoreticalMaximumSpeedup)
		row.WriteString(strconv.FormatInt(int64(numberOfProcesses), 10) + "," + strconv.FormatFloat(achievedSpeedup, 'f', -1, 64) + "," + strconv.FormatInt(int64(theoreticalMaximumSpeedup), 10) + "\r\n")
	}
	exportScalingFile(row.String(), false)
	fmt.Println("End weak scaling.")

}

func main() {
	/*
		numberOfSimulationsSerial := 10
		predictionWindowSizeSerial := 100
		numberOfProcessesSerial := 1
		monteCarloSimulationFinanceSerial := MonteCarloSimulationFinance{
			tickerSymbol:      "AAPL",
			startDate:         "2000-01-01",
			endDate:           "2020-01-01",
			numberOfProcesses: numberOfProcessesSerial}
		monteCarloSimulationFinanceSerial.dataAcquisition()
		monteCarloSimulationFinanceSerial.calculatePeriodicDailyReturn()
		fmt.Println("Stock market price predictions using the Monte Carlo simulation serial version")
		serialPredictions, serialExecutionTime := monteCarloSimulationFinanceSerial.mcsFinanceSerial(numberOfSimulationsSerial, predictionWindowSizeSerial)
		fmt.Printf("Execution time(n = %d, p = %d, w = %d) = %f seconds\r\n",
			numberOfSimulationsSerial, numberOfProcessesSerial,
			predictionWindowSizeSerial, serialExecutionTime)
		monteCarloSimulationFinanceSerial.exportFinanceFileSerial(serialPredictions)
	*/
	/*
		numberOfSimulationsParallel := 10
		predictionWindowSizeParallel := 100
		numberOfProcessesParallel := 4
		monteCarloSimulationFinanceParallel := MonteCarloSimulationFinance{
			tickerSymbol:      "AAPL",
			startDate:         "2000-01-01",
			endDate:           "2020-01-01",
			numberOfProcesses: numberOfProcessesParallel}
		monteCarloSimulationFinanceParallel.dataAcquisition()
		monteCarloSimulationFinanceParallel.calculatePeriodicDailyReturn()
		fmt.Println("Stock market price predictions using the Monte Carlo simulation parallel version")
		parallelPredictions, parallelExecutionTime := monteCarloSimulationFinanceParallel.mcsFinanceParallel(numberOfSimulationsParallel, predictionWindowSizeParallel)

		fmt.Printf("Execution time(n = %d, p = %d, w = %d) = %f seconds\r\n",
			numberOfSimulationsParallel, numberOfProcessesParallel,
			predictionWindowSizeParallel, parallelExecutionTime)

		monteCarloSimulationFinanceParallel.exportFinanceFileParallel(parallelPredictions)
	*/
	strongScaling()
	weakScaling()

}

File: Python/MonteCarloSimulationFinance/MonteCarloSimulationFinance.py
import copy
import math
import random
import time
from multiprocessing import Pool

import numpy as np
from pandas_datareader import data
from scipy.stats import norm


def calculate_execution_time(function):
    def calculate_duration(*args, **kwargs):
        start_time = time.time()
        executing_function = function(*args, **kwargs)
        end_time = time.time()
        execution_time = round(end_time - start_time, 7)
        return executing_function, execution_time

    return calculate_duration


class MonteCarloSimulationFinance:
    def __init__(self, start_date, end_date, ticker_symbol, number_of_processes):
        self.start_date = start_date
        self.end_date = end_date
        self.ticker_symbol = ticker_symbol
        self.data = None
        self.time_series = None
        self.number_of_processes = number_of_processes
        self.parallel_flag = False

    def data_acquisition(self):
        stock = data.DataReader(self.ticker_symbol, 'yahoo', self.start_date, self.end_date)
        stock = stock.dropna()
        self.time_series = stock['Close']

    # Differencing time series = Shifting and lagging time series
    def calculate_periodic_daily_return(self):
        # Differencing time series
        # The diff() function calculates the first differences of the time series.
        self.data = np.log(self.time_series).diff().dropna()
        # Shifting and lagging time series
        # self.data=np.log(self.time_series / self.time_series.shift(1)).dropna()

    # A z-table, also called the standard normal table, is a mathematical table that allows us to know
    # the percentage of values below (to the left) a z-score in a standard normal distribution (SND).
    # A z-score, also known as a standard score, indicates the number of standard deviations
    # a raw score lays above or below the mean. When the mean of the z-score is calculated it is always 0,
    # and the standard deviation (variance) is always in increments of 1.
    def calculate_z_score(self):
        return norm.ppf(random.random())

    def calculate_average_daily_return(self):
        return np.mean(self.data)

    def calculate_variance(self):
        return np.var(self.data)

    def calculate_standard_deviation(self):
        return np.std(self.data)

    def calculate_drift(self):
        return self.calculate_average_daily_return() - self.calculate_variance() / 2

    def calculate_random_value(self):
        return self.calculate_standard_deviation() * self.calculate_z_score()

    # prediction window size: number of prediction days per simulation
    def simulation_finance(self, number_of_simulations, prediction_window_size):
        predictions = []
        prediction = []
        for i in range(number_of_simulations):
            # today’s price
            prediction.append(self.time_series.iloc[-1])
            for j in range(prediction_window_size):
                # Next Day’s Price=Today’s Price × e^(Drift+Random Value)
                prediction.append(
                    prediction[-1] * pow(math.e, (self.calculate_drift() + self.calculate_random_value())))

            predictions.append(copy.deepcopy(prediction))
            prediction.clear()
        return predictions

    @calculate_execution_time
    def mcs_finance_serial(self, number_of_simulations, prediction_window_size):
        self.parallel_flag = False
        return self.simulation_finance(number_of_simulations, prediction_window_size)

    @calculate_execution_time
    def mcs_finance_parallel(self, number_of_simulations, prediction_window_size):
        self.parallel_flag = True
        pool = Pool(processes=self.number_of_processes)
        number_of_simulations_per_process = int(number_of_simulations / self.number_of_processes)
        simulations_per_process = []
        # Append the same value multiple times to a list
        # To add v, n times, to l:
        # l += n * [v]
        # Mapping a function with multiple arguments to a multiprocessing pool will distribute
        # the input data across processes to be run with the referenced function.
        simulations_per_process += self.number_of_processes * [
            (number_of_simulations_per_process, prediction_window_size)]
        predictions = pool.starmap(self.simulation_finance, simulations_per_process)
        return predictions

    def export_finance_file(self, predictions):
        serial_number = 1
        if self.parallel_flag == False:
            # r before string converts normal string to raw string
            path = r"C:\Users\Dule\Desktop\NAPREDNE TEHNIKE PROGRAMIRANJA\PROJEKAT\NTP" \
                   r"\Execution Results\Finance\PythonFinanceSerial.txt"
            out_file = open(path, "w")
            for i in range(len(predictions)):
                for j in range(len(predictions[i])):
                    if j == 0:
                        # Write a serial number of simulations at the beginning of a file.
                        out_file.write(str(serial_number) + "," + " ")
                    # A component in one row of the output file.
                    out_file.write(str(predictions[i][j]))
                    if j == (len(predictions[i]) - 1):
                        out_file.write("\r\n")
                        serial_number += 1
                    else:
                        out_file.write("," + " ")

            out_file.close()

        else:
            # r before string converts normal string to raw string
            path = r"C:\Users\Dule\Desktop\NAPREDNE TEHNIKE PROGRAMIRANJA\PROJEKAT\NTP" \
                   r"\Execution Results\Finance\PythonFinanceParallel.txt"
            out_file = open(path, "w")
            for i in range(len(predictions)):
                for j in range(len(predictions[i])):
                    for k in range(len(predictions[i][j])):
                        if k == 0:
                            # Write a serial number of simulations at the beginning of a file.
                            out_file.write(str(serial_number) + "," + " ")
                        # A component in one row of the output file.
                        out_file.write(str(predictions[i][j][k]))
                        if k == (len(predictions[i][j]) - 1):
                            out_file.write("\r\n")
                            serial_number += 1
                        else:
                            out_file.write("," + " ")

            out_file.close()


if __name__ == "__main__":
    number_of_simulations_serial = 10
    prediction_window_size_serial = 100
    number_of_processes_serial = 1
    monte_carlo_simulation_finance_serial = MonteCarloSimulationFinance(
        '1980-01-01', '2019-12-31', 'AAPL', number_of_processes_serial)
    monte_carlo_simulation_finance_serial.data_acquisition()
    monte_carlo_simulation_finance_serial.calculate_periodic_daily_return()
    serial_predictions, serial_execution_time = monte_carlo_simulation_finance_serial.mcs_finance_serial(number_of_simulations_serial,
                                                                                                  prediction_window_size_serial)
    print("Stock market price predictions using the Monte Carlo simulation serial version")
    print("Execution time(n = {}, p = {}, w = {}) = {} seconds".format(number_of_simulations_serial,
                                                                       number_of_processes_serial,
                                                                       prediction_window_size_serial, serial_execution_time))
    monte_carlo_simulation_finance_serial.export_finance_file(serial_predictions)

    number_of_simulations_parallel = 10
    prediction_window_size_parallel = 100
    number_of_processes_parallel = 4
    monte_carlo_simulation_finance_parallel = MonteCarloSimulationFinance('1980-01-01', '2019-12-31', 'AAPL',
                                                                 number_of_processes_parallel)
    monte_carlo_simulation_finance_parallel.data_acquisition()
    monte_carlo_simulation_finance_parallel.calculate_periodic_daily_return()
    parallel_predictions, parallel_execution_time = monte_carlo_simulation_finance_parallel.mcs_finance_parallel(
        number_of_simulations_parallel,
        prediction_window_size_parallel)
    print("Stock market price predictions using the Monte Carlo simulation parallel version")
    print("Execution time(n = {}, p = {}, w = {}) = {} seconds".format(number_of_simulations_parallel,
                                                                       number_of_processes_parallel,
                                                                       prediction_window_size_parallel, parallel_execution_time))
    monte_carlo_simulation_finance_parallel.export_finance_file(parallel_predictions)



















